core/src/main/java/org/infinispan/commands/write/PutMapCommand.java

```diff
@@ -92,13 +92,26 @@ public Object perform(InvocationContext ctx) throws Throwable {
       Map<Object, Object> previousValues = new HashMap<>();
       for (Entry<Object, Object> e : map.entrySet()) {
          Object key = e.getKey();
-         MVCCEntry me = lookupMvccEntry(ctx, key);
-         if (me != null) {
-            Object value = me.getValue();
-            previousValues.put(key, value);
-            notifier.notifyCacheEntryModified(key, e.getValue(), metadata, value, me.getMetadata(), true, ctx, this);
-            me.setValue(e.getValue());
-            me.setChanged(true);
+         MVCCEntry contextEntry = lookupMvccEntry(ctx, key);
+         if (contextEntry != null) {
+            Object newValue = e.getValue();
+            Object previousValue = contextEntry.getValue();
+            Metadata previousMetadata = contextEntry.getMetadata();
+
+            // Even though putAll() returns void, QueryInterceptor reads the previous values
+            // TODO The previous values are not correct if the entries exist only in a store
+            previousValues.put(key, previousValue);
+
+            if (contextEntry.isCreated()) {
+               notifier.notifyCacheEntryCreated(key, newValue, metadata, true, ctx, this);
+            }
+            // TODO Only invoke the modified listener if the entry was not created?
+            notifier.notifyCacheEntryModified(key, newValue, metadata, previousValue, previousMetadata, true,
```

Comment:
Unfortunately this conversation got removed https://github.com/infinispan/infinispan/pull/3705#discussion_r39600752